runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -1342,6 +1351,14 @@ private void onRemovedEventTimer(TimerData removedTimer) {
       }
     }
 
+    private void onRemovedSdfTimer(TimerData removedTimer) {
+      Preconditions.checkState(
+          StateAndTimerBundleCheckpointHandler.isSdfTimer(removedTimer.getTimerId()));
+      Preconditions.checkState(timerUsesOutputTimestamp(removedTimer));
+      // Remove the watermark hold which is set for this sdf timer.
+      keyedStateInternals.removeWatermarkHoldUsage(removedTimer.getOutputTimestamp());
+    }
```

Comment:
Wouldn't it make sense to integrate this check with the `timerUsesOutputTimestamp` method?

Reply:
We could rename the mentioned method to onFiredTimer and include the checks for output watermark holds in there.